src/libstd/thread/local.rs

```diff
@@ -130,31 +130,42 @@ impl<T: 'static> fmt::Debug for LocalKey<T> {
 #[stable(feature = "rust1", since = "1.0.0")]
 #[allow_internal_unstable]
 macro_rules! thread_local {
-    // rule 0: empty (base case for the recursion)
+    // empty (base case for the recursion)
     () => {};
 
-    // rule 1: process multiple declarations where the first one is private
+    // process multiple declarations where the first one is private
     ($(#[$attr:meta])* static $name:ident: $t:ty = $init:expr; $($rest:tt)*) => (
-        thread_local!($(#[$attr])* static $name: $t = $init); // go to rule 2
+        __thread_local_inner!($(#[$attr])* [] $name, $t, $init);
         thread_local!($($rest)*);
     );
-
-    // rule 2: handle a single private declaration
+    
+    // handle a single private declaration
     ($(#[$attr:meta])* static $name:ident: $t:ty = $init:expr) => (
-        $(#[$attr])* static $name: $crate::thread::LocalKey<$t> =
-            __thread_local_inner!($t, $init);
+        __thread_local_inner!($(#[$attr])* [] $name, $t, $init);
     );
 
-    // rule 3: handle multiple declarations where the first one is public
+    // handle multiple declarations where the first one is public
     ($(#[$attr:meta])* pub static $name:ident: $t:ty = $init:expr; $($rest:tt)*) => (
-        thread_local!($(#[$attr])* pub static $name: $t = $init); // go to rule 4
+        __thread_local_inner!($(#[$attr])* [pub] $name, $t, $init);
         thread_local!($($rest)*);
     );
-
-    // rule 4: handle a single public declaration
+    
+    // handle multiple public declarations
     ($(#[$attr:meta])* pub static $name:ident: $t:ty = $init:expr) => (
-        $(#[$attr])* pub static $name: $crate::thread::LocalKey<$t> =
-            __thread_local_inner!($t, $init);
+        __thread_local_inner!($(#[$attr])* [pub] $name, $t, $init);
+        thread_local!($($rest)*);
+    );
+    
+    // handle multiple declarations where the first one is restricted public
+    ($(#[$attr:meta])* pub $vis:tt static $name:ident: $t:ty = $init:expr; $($rest:tt)*) => (
+        __thread_local_inner!($(#[$attr])* [pub $vis] $name, $t, $init);
+        thread_local!($($rest)*);
+    );
+    
+    // rule 4: handle multiple restricted public declarations
+    ($(#[$attr:meta])* pub $vis:tt static $name:ident: $t:ty = $init:expr) => (
+        __thread_local_inner!($(#[$attr])* [pub $vis] $name, $t, $init);
+        thread_local!($($rest)*);
```

Comment:
Please note these six rules can be reduced to four if we revive [the `$x:vis` RFC](https://github.com/rust-lang/rfcs/pull/1575).